typing/typedecl.ml

```diff
@@ -2316,11 +2316,24 @@ let check_well_founded ~abs_env env loc path to_check visited ty0 =
     match get_desc ty with
     | Tconstr(p, tyl, _) ->
         let to_check = to_check p in
-        if to_check then List.iter (check_subtype parents trace ty) tyl;
+        let check_parameters () =
+          if tyl <> [] then begin
+            let variance = 
+              try (Env.find_type p env).type_variance
+              with Not_found -> List.map (fun _ -> Variance.unknown) tyl
+            in
+            List.iter2 
+              (fun v t ->
+                 if Variance.(mem May_noncontractive v) then
+                   check_subtype parents trace ty t)
+              variance tyl
+          end
+        in
+        if to_check then check_parameters ();
```

Comment:
I don't think this is wrong, but I'm not sure if I'd be able to tell whether it is. It looks right, though.